lib/improver/tests/spotdata/spotdata/test_site_data.py

```diff
@@ -0,0 +1,300 @@
+            site.pop('longitude')
+        self.save_json(self.site_data)
+
+        msg = 'longitude and latitude must be defined'
+        with self.assertRaisesRegexp(KeyError, msg):
+            Plugin(self.method).process(self.site_path)
+
+    def test_unequal_lat_lon(self):
+        """
+        Test loading a valid json file containing site details,
+        but missing compulsary longitude data for some sites.
+
+        """
+        self.site_data[0].pop('longitude')
+        self.save_json(self.site_data)
+        expected = 4  # No. of sites expected.
+        result = Plugin(self.method).process(self.site_path)
+
+        self.assertEqual(len(result), expected)
+        self.assertNotIn(16, np.array([index for index in result]))
+
+    def test_only_lat_lon(self):
+        """
+        Test loading a valid json file containing site details, only containing
+        lat/lon. Altitudes set to fill value of -9999, and utc_offset
+        calculated from longitude.
+
+        """
+        for site in self.site_data:
+            site.pop('bestdata_id')
+            site.pop('altitude')
+            site.pop('gmtoffset')
+
+        self.save_json(self.site_data)
+        result = Plugin(self.method).process(self.site_path)
+        expected_altitudes = [-9999.] * 5
+        expected_utc_offsets = [-4, -5, -1, 1, 10]
+
+        self.assertArrayEqual(
+            expected_altitudes,
+            np.array([result[index]['altitude'] for index in result]))
+        self.assertArrayEqual(
+            expected_utc_offsets,
+            np.array([result[index]['utc_offset'] for index in result]))
+
+
+class Test_runtime_list(Test_ImportSiteData):
+    """
+    Test setting up site data at run time through the CLI.
+
+    """
+    method = 'runtime_list'
+
+    def test_complete(self):
+        """
+        Test creating sites at runtime. Here only latitude and longitude are
+        provided.
+
+        """
+        altitudes = range(5)
+        site_ids = [10, 20, 30, 40, 50]
+        result = Plugin(self.method).process(self.latitudes, self.longitudes,
+                                             altitudes=altitudes,
+                                             site_ids=site_ids)
+        expected_utc_offsets = [-4, -5, -1, 1, 10]
+
+        self.assertArrayEqual(
+            self.latitudes,
+            np.array([result[index]['latitude'] for index in result]))
+        self.assertArrayEqual(
+            self.longitudes,
+            np.array([result[index]['longitude'] for index in result]))
+        self.assertArrayEqual(
+            altitudes,
+            np.array([result[index]['altitude'] for index in result]))
+        self.assertArrayEqual(
+            expected_utc_offsets,
+            np.array([result[index]['utc_offset'] for index in result]))
+
+        # bestdata_id used as key in resulting dictionary.
+        self.assertArrayEqual(site_ids,
+                              np.array([index for index in result]))
+
```

Comment:
If you did change to a list of dictionary you would only need to check most of these things (e.g. unequal lat lon) once for the main processing function and that you have been supplied by a list of dictionaries for runtime_list method and a json file for from_file method.